File: aoc/2018/day11/part_2.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int calc_power(int x, int y, int serial) {
	int rack = x+10;
	int power = rack*y;
	power += serial;
	power *= rack;
	power = power / 100 % 10;
	power -= 5;
	return power;
}

const int SIZE = 301;
const int SERIAL = 5235;

void run() {
	int* grid[SIZE][SIZE];

	/* base case */
	for(int x = 1; x < SIZE; ++x) {
		for(int y = 1; y < SIZE; ++y){
			grid[x][y] = malloc(sizeof(int) * (SIZE - x + 1));
			grid[x][y][0] = calc_power(x, y, SERIAL);
		}
	}

	int max_power = 0;

	/* case 1 -> 300 */
	for (int r = 1; r < SIZE; ++r){
		for(int x = 1; x < SIZE - r; ++x){
			for(int y = 1; y < SIZE - r; ++y){
				int power = (
						  grid[x + 1][y    ][r - 1] /* top right */
						+ grid[x    ][y + 1][r - 1] /* bottom left */
						+ grid[x    ][y    ][0]     /* top left 1 square */
						+ grid[x + r - 1][y + r - 1][0] /* bottom right 1 square */
					);
				if (r > 1) {
						/* don't need to subtract for case #1 */
						power -= grid[x + 1][y + 1][r - 2]; /* uncount middle */
				}
				grid[x][y][r] = power;

				if(power > max_power) {
					max_power = power;
					printf("%d,%d,%d\t%d\n", x, y, r, power);
				}
			}
		}
	}

	for(int x=1; x<SIZE; ++x){
		for(int y =1; y<SIZE;++y){
			free(grid[x][y]);
			grid[x][y] = NULL;
		}
	}
}

int main() {
	clock_t start = clock();
	run();
	clock_t end = clock();
	printf("time = %.3f ms\n", ((float)(end - start)) / CLOCKS_PER_SEC * 1000);

	return 0;
}

File: aoc/2018/day11/part_2.py
# from itertools import range

def calc_power(x, y, serial):
    rack = x + 10
    power = rack * y
    power += serial
    power *= rack
    if power < 100:
        power = 0
    else:
        power = int(str(power)[-3])
    power -= 5
    return power

max_v = 0
max_x=0
max_y=0
max_r=0

n=301
grid = [[[0 for k in range(n)] for j in range(n)] for i in range(n)]

#base
for x in range(1,301):
    for y in range(1,301):
        grid[x][y][0] = calc_power(x, y, 5235)

for r in range(1, 301):
    # print(r)
    for x in range(1,300-r+1):
        for y in range(1,300-r+1):
            a = (
                   grid[x + 1][y    ][r - 1]  #top right
                 + grid[x    ][y + 1][r - 1] #top left
                 - grid[x + 1][y + 1][r - 2] #less center
                 + grid[x    ][y    ][0    ] #top left point
                 + grid[x + r - 1][y + r - 1][0]) #bot right point
            grid[x][y][r] = a
            # print(a)
            if a > max_v:
                max_v = a
                max_x = x
                max_y = y
                max_r = r

                print("{},{},{}\t{}".format(max_x,max_y,max_r, a))
